src/qibojit/custom_operators/backends.py

```diff
@@ -384,3 +383,281 @@ def swap_pieces(self, piece0, piece1, new_global, nlocal):
+            gate_ptr = gate.data.ptr
+        elif isinstance(gate, self.np.ndarray):
+            gate_ptr = gate.ctypes.data
+        else:
+            raise ValueError
+
+        workspaceSize = self.cusv.apply_matrix_buffer_size(handle,
+                                                           data_type,
+                                                           nqubits,
+                                                           gate_ptr,
+                                                           data_type,
+                                                           self.cusv.MatrixLayout.ROW,
+                                                           adjoint,
+                                                           ntarget,
+                                                           ncontrols,
+                                                           compute_type
+                                                           )
+
+        # check the size of external workspace
+        if workspaceSize > 0:
+            workspace = self.cp.cuda.memory.alloc(workspaceSize)
+            workspace_ptr = workspace.ptr
+        else:
+            workspace_ptr = 0
+
+        self.cusv.apply_matrix(handle,
+                               state.data.ptr,
+                               data_type,
+                               nqubits,
+                               gate_ptr,
+                               data_type,
+                               self.cusv.MatrixLayout.ROW,
+                               adjoint,
+                               target.ctypes.data,
+                               ntarget,
+                               controls.ctypes.data,
+                               ncontrols,
+                               0,
+                               compute_type,
+                               workspace_ptr,
+                               workspaceSize
+                               )
+
+        self.cusv.destroy(handle)
+        return state
+
+    def collapse_state(self, state, qubits, result, nqubits, normalize=True):
+
+        handle = self.cusv.create()
+        state = self.cast(state)
+        results = bin(result).replace("0b", "")
+        results = list(map(int,  '0'* (len(qubits) - len(results)) + results))[::-1]
+        ntarget = 1
+        qubits = self.np.asarray(qubits.get(), dtype = self.np.int32)
+        data_type, compute_type = self.get_cuda_type(state.dtype)
+
+        for i  in range(len(results)):
+            self.cusv.collapse_on_z_basis(handle,
+                                          state.data.ptr,
+                                          data_type,
+                                          nqubits,
+                                          results[i],
+                                          [qubits[i]],
+                                          ntarget,
+                                          1
+                                          )
+
+
+        if normalize:
```

Comment:
If I remove this if statement tests on qibojit don't pass for me. 
It's not very clear from their documentation. The norm parameter is defined as follows:
```
norm (double) – The normalization factor for the statevector after collapse.
```
A did a few tests and it seems like the state is just multiply by 1/sqrt(norm) after the collapse.
This is confusing because in order to obtain a normalized state you will need to compute the norm and the pass it as a parameter. However you cannot compute the correct norm if you have not collapse the state yet.